text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1160 @@
+attributes or labels or otherwise have a certain shape and composition of data
+(this is referred to as "telemetry schema" throughout this document).
+
+This makes it difficult or impossible to change the composition of the emitted
+telemetry data without breaking the consumers. For example changing the name of
+an attribute of a span created by an instrumentation library can break the
+backend if the backend expects to find that attribute by its name.
+
+Semantic conventions are an important part of this problem. These conventions
+define what names and values to use for span attributes, metric names and other
+fields. If semantic conventions are changed the existing implementations
+(telemetry source or consumers) need to be also changed correspondingly.
+Furthermore, to make things worse, the implementations of telemetry sources and
+implementations of telemetry consumers that work together and that depend on the
+changed semantic convention need to be changed simultaneously, otherwise such
+implementations will no longer work correctly together.
+
+Essentially there is a coupling between 3 parties: 1) OpenTelemetry semantic
+conventions, 2) telemetry sources and 3) telemetry consumers. The coupling
+complicates the independent evolution of these 3 parties.
+
+We recognize the following needs:
+
+- OpenTelemetry semantic conventions need to evolve over time. When conventions
+  are first defined, mistakes are possible and we may want to fix the mistakes
+  over time. We may also want to change conventions to re-group the attributes
+  into different namespaces as our understanding of the attribute taxonomy
+  improves.
+
+- Telemetry sources over time may want to change the schema of the telemetry
+  they emit. This may be because for example the semantic conventions evolved
+  and we want to make our telemetry match the newly introduced conventions.
+
+- In an observability system there may simultaneously exist telemetry sources
+  that produce data that conforms to different telemetry schemas because
+  different sources evolve at a different pace and are implemented and
+  controlled by different entities.
+
+- Telemetry consumers have a need to understand what schema a particular piece
+  of received telemetry confirms to. The consumers also need a way to be able to
+  interpret the telemetry data that uses different telemetry schemas.
+
+This document proposes a solution to these needs.
+
+## Solution Summary
+
+We believe that the 3 parties described above should be able to evolve
+independently over time, while continuously retaining the ability to correctly
+work together.
+
+Telemetry Schemas are central to how we make this possible. Here is a summary of
+the proposal:
+
+- We introduce a file format for defining telemetry schemas.
+
+- Telemetry schemas are versioned. Over time the schema may evolve and telemetry
+  sources may emit data confirming to newer versions of the schema.
```

Comment:
Should Telemetry Schema versions correspond to versions of the OpenTelemetry spec? Or would those be two different (but linked) lines of versions?

Reply:
There is no direct match between Otel spec version numbers and Otel schema version numbers. There is no need to have such coupling, the two should be able evolve independently. 
Yes, a particular Otel spec X version will naturally correspond a particular Otel schema version Y (since semantic conventions are part of the spec and schemas are tied to semantic conventions), however X may and likely will be different from Y.

Reply:
This feels rather confusing to me. What's the argument against keeping schema and spec version number the same?

Reply:
If we want the version numbers the same then we have to align releases of the spec with changes in schema. I was thinking that keeping these decoupled gives us more freedom to evolve the 2 independently.

Not a strong opinion though. I think using the same version numbers would work too. It just means that in some cases we will release a new schema version when there are no changes to it (because it is time to release a new spec version but nothing is changed in semantic convention). That's fine, we just need to be careful not to forget to keep schema version number in sync with spec version number (an automated pre-release check would be preferable).
Likewise, sometimes there is probably nothing new in the spec other than changes to semantic conventions, so we will have to release a new spec version to get a new version number we can use for new schema.

If we feel keeping version numbers in sync is less confusing, then I don't mind it, that can work.

Reply:
I think requiring a spec release for an updated schema would be a very good thing.

Reply:
I'd be in favor of sharing the same version numbers. I expect people writing instrumentation according to a certain version of the specification. It seems natural to me that this version is the same as the schema version.

Reply:
The version number recommendations section I have in the doc will no longer work in that case. Let me think about this.

Reply:
Sharing them means we will have as many schema versions as API ones, which feels odd and I don't remember seeing a schema with that many versions.

Reply:
I spent a bit time on this and I believe I agree that it is preferable to keep the schema version numbers in sync with the spec version numbers.

@carlosalberto I understand your concern. In some cases there will be no actual changes in the schema, but we will still bump the version number to keep it in sync with spec version number. It will mean adding one line to the schema YAML file. We can also automatically verify that this line is correctly added, so I think it is not a big problem and the trade-off is worth it to avoid confusion and the need to explain mismatching version numbers.
I will modify the proposal to reflect this.

Reply:
I changes sections "Schema Version Number" and "OpenTelemetry Schema", please have a look.

Reply:
Thanks. I think it will make life easier for people writing instrumentation: they need to keep track of only two versions (OpenTelemetry spec and language specific API) instead of three (spec, language API, and telemetry schema).

Reply:
@yurishkuro @carlosalberto are you OK with the changes in "Schema Version Number" and "OpenTelemetry Schema" sections?

Reply:
@tigrannajaryan I'm fine with it. We will have to validate it once this hits the Specification + implementation, so we can still revert if truly needed.